---
title: "MixN_ChronicFetal"
output:
    html_document:
        toc: true
        toc_float: true
        collapsed: true
        theme: 'simplex'
        highlight: 'espresso'
        code_folding: hide
---

### Data loading

```{r, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SEtools)
  library(edgeR)
  library(DT)
  library(pheatmap)
  library(plotly)
  library(dplyr)
  library(sva)
  library(reshape2)
  library(overlapper)
  library(ggrepel)
})
source("Functions/EDC_Functions.R")
source("Functions/CriFormatted.R")
load("Data/AllSEcorrected.RData",verbose=T)
```

### DEGs heatmaps
```{r,message=FALSE,collapse=TRUE, fig.width=10, fig.height=10}
fetal.chronic <- row.names(DEAs$chronic.fetal)[which((abs(DEAs$chronic.fetal$logFC.EXPO1X)>0.5 | abs(DEAs$chronic.fetal$logFC.EXPO1000X)>0.5) & DEAs$chronic.fetal$FDR<=0.05 & DEAs$chronic.fetal$logCPM>0)]


MixN_fetal_chronic <- SEs$chronic.fetal[,which(SEs$chronic.fetal$EXPO=="CNT"|SEs$chronic.fetal$EXPO=="1X" | SEs$chronic.fetal$EXPO=="1000X")]
sehm(MixN_fetal_chronic, assayName = "corrected", fetal.chronic,  do.scale = T, show_rownames = F, anno_columns = c("EXPO2"), main="Chronic fetal DEGs")
```


### 1X vs 1000X


```{r,message=FALSE,collapse=TRUE, fig.height=7.5, fig.width=7.5}
FDRTh <- 0.05
LogFCTh <- 0.5

MixN1X <- DEAs$chronic.fetal
MixN1X$genes <- row.names(MixN1X)
MixN1X$logFC <- (MixN1X$logFC.EXPO1X)

MixN1000X <- DEAs$chronic.fetal
MixN1000X$genes <- row.names(MixN1000X)
MixN1000X$logFC <- (MixN1X$logFC.EXPO1000X)

```

```{r,message=FALSE,collapse=TRUE, fig.height=7.5, fig.width=7.5}
Comp <- compareResultsFCNew(MixN1X, MixN1000X, FDRth=0.05, FCth=2^0.5, FDRceil=1e-10, logCPMth=0, FCceil=2.5, 
                              title='Comparing MixN 1X to 1000X', geneLabel=TRUE, topLab=-16)

Comp$Scatter

ggsave(Comp$Scatter, filename='Data/ChronicFetal/Scatter.pdf', width=10, height=10)
ggsave(Comp$Scatter, filename='Data/ChronicFetal/Scatter.png', width=10, height=10)
```



### Neurodevelopmental Disorder Genes 

```{r, message=FALSE,collapse=TRUE, fig.height=6}
load("Data/ASD.RData", verbose = T)

fetal.chronic3 <- row.names(DEAs$chronic.fetal)[which(DEAs$chronic.fetal$FDR<=0.05)]
controlNeg <- row.names(DEAs$chronic.fetal)[-which(DEAs$chronic.fetal$FDR<=0.05)]

#random1 <- sample(row.names(DEAs$chronic.fetal), 10)
#random2 <- sample(row.names(DEAs$chronic.fetal), 100)
#random3 <- sample(row.names(DEAs$chronic.fetal), 1000)

fetal.chronicUp <- row.names(DEAs$chronic.fetal)[which(((DEAs$chronic.fetal$logFC.EXPO1X)>0 & (DEAs$chronic.fetal$logFC.EXPO1000X)>0) & DEAs$chronic.fetal$FDR<=0.05)]
fetal.chronicDown <- row.names(DEAs$chronic.fetal)[which(((DEAs$chronic.fetal$logFC.EXPO1X)<0 & (DEAs$chronic.fetal$logFC.EXPO1000X)<0) & DEAs$chronic.fetal$FDR<=0.05)]


#MixN <- list(fetalDEGs=fetal.chronic3, fetalDEGsUp=fetal.chronicUp, fetalDEGsDown=fetal.chronicDown,NonAffectedGenes=controlNeg)
MixN <- list(fetalDEGs=fetal.chronic3, NonAffectedGenes=controlNeg)

m <- overlapper::multintersect(ll = MixN, ll2 = ASD, universe = row.names(DEAs$chronic.fetal),two.tailed = F)

fetal.chronicLong=fetal.chronic3
save(fetal.chronic,fetal.chronicLong,file = "Data/DEGsFetalChronic.RData")
```


```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, sizeRange = c(0,15), th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicFetal/DotplotFetalNDD.pdf', width=10, height=6)

```

#### P values of the overlaps 

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m$prob <- round(m$prob, digits=3)
DT::datatable(m$prob)
```



### NDD psychencode
```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m <- overlapper::multintersect(ll = MixN, ll2 = PsychencodeNDD, universe = row.names(DEAs$chronic.fetal),two.tailed = F)
```

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicFetal/DotplotFetalNDD.pdf', width=10, height=6)
```





### Gene expression dysregulation of the most important NDD genes

```{r, message=FALSE,collapse=TRUE}
geneStripPairEDCMix(SE = MixN_fetal_chronic,GeneSet = intersect(fetal.chronic, unlist(SFARIgenes)), printExp = FALSE, SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.fetal[intersect(fetal.chronic, unlist(SFARIgenes)),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`
 


### Gene expression dysregulation of the chronic organoids NDD-DEGs in fetal dataset

```{r, message=FALSE,collapse=TRUE}
load("Data/DEGsOrganoidsChronic.RData", verbose = T)
geneStripPairEDCMix(SE = MixN_fetal_chronic,GeneSet = intersect(org.chronic, union(union(SFARIgenes$score1,SFARIgenes$score2),SFARIgenes$score3)), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.fetal[intersect(org.chronic, union(union(SFARIgenes$score1,SFARIgenes$score2),SFARIgenes$score3)),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`
 

### Gene expression dysregulation of the genes found in the acute expo

```{r, message=FALSE,collapse=TRUE}
load("Data/DEGsFetalAcute.RData", verbose = T)
geneStripPairEDCMix(SE = MixN_fetal_chronic,GeneSet = intersect(fet.acute,fetal.chronic), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.fetal[intersect(fet.acute,fetal.chronic),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`


```{r, message=FALSE,collapse=TRUE}
geneStripPairEDCMix(SE = MixN_fetal_chronic,GeneSet = c("CLSTN2", "EPHB2"), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.fetal[c("CLSTN2", "EPHB2"),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`

### Data praparation

For details on data filtering, normalization, batch correction and differential expression analysis, see [here](./DataPreparation/SVAcorrectedObjects.html) 

For details on the preparation of Neurodevelopmental disorder genes, see [here](./DataPreparation/ASD/DataPreparation.html) 

***

```{r child='footer.Rmd'}
```

```{r SaveSession}
sessionInfo()
```
