---
title: "MixN_ChronicOrganoids"
output:
    html_document:
        toc: true
        toc_float: true
        collapsed: true
        theme: 'simplex'
        highlight: 'espresso'
        code_folding: hide
---

### Data loading

```{r, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SEtools)
  library(edgeR)
  library(DT)
  library(pheatmap)
  library(plotly)
  library(dplyr)
  library(sva)
  library(reshape2)
  library(overlapper)
  library(ggrepel)
})
source("Functions/EDC_Functions.R")
source("Functions/CriFormatted.R")
load("Data/TotalSumExp.RData")
load("Data/geneLengths.RData")
load("Data/AllSEcorrected.RData",verbose=T)
load("Data/brainspan.cortex.byAge.RData", verbose = T)
load("Data/iPSC.RData", verbose = T)
```

### Brainspan correlation
```{r,message=FALSE,collapse=TRUE, fig.width=10, fig.height=10}
Organoids <- assays(SEs$chronic.org[,which(SEs$chronic.org$EXPO=="CNT")])$counts
Fetal2D <- assays(SEs$chronic.fetal[,which(SEs$chronic.fetal$EXPO=="CNT")])$counts
iPSC <- e[,grep("3391B",colnames(e))]
Fetal <- byAge
i <- intersect(rownames(Organoids),intersect(rownames(iPSC),rownames(Fetal2D)))
tot <- cbind(Organoids[i,],iPSC[i,],Fetal2D[i,])

tot <- filterGenes(tot)
en <- donorm(tot)
#put the rows in the same order
i <- intersect(row.names(Fetal),row.names(en))
en <- en[i,]
Fetal <- Fetal[i,]
#create the matrix of pearson correlation coefficients and plot
m <- matrix(0,nrow=ncol(Fetal),ncol=ncol(en))
for(i in 1:ncol(en)){ m[,i] <- apply(Fetal,2,FUN=function(x){ cor(log(en[,i]+1),log(x+1)) }) }
row.names(m) <- colnames(Fetal)
colnames(m) <- colnames(en)
m2 <- m[grep("pcw",row.names(m)),]
colnames(m2) <- c(rep("Organoids",8), rep("iPSC",5), rep("FetalProgenitors",8))
byheatmap(m2,cluster_row=F)
```

### DEGs heatmaps
```{r,message=FALSE,collapse=TRUE, fig.width=10, fig.height=10}
org.chronic <- row.names(DEAs$chronic.org)[which((abs(DEAs$chronic.org$logFC.EXPO1X)>0.5 | abs(DEAs$chronic.org$logFC.EXPO1000X)>0.5) & DEAs$chronic.org$FDR<=0.05 & DEAs$chronic.org$logCPM > 0)]

MixN_org_chronic <- SEs$chronic.org[,which(SEs$chronic.org$EXPO=="CNT"|SEs$chronic.org$EXPO=="1X" | SEs$chronic.org$EXPO=="1000X")]

sehm(MixN_org_chronic, assayName = "corrected", org.chronic,  do.scale = T, show_rownames = F, anno_columns = c("Line","EXPO2"), main="Chronic org DEGs")

```



### 1X vs 1000X

```{r,message=FALSE,collapse=TRUE, fig.width=10, fig.height=10}
FDRTh <- 0.05
LogFCTh <- 0.5

MixN1X <- DEAs$chronic.org
MixN1X$genes <- row.names(MixN1X)
MixN1X$logFC <- (MixN1X$logFC.EXPO1X)

MixN1000X <- DEAs$chronic.org
MixN1000X$genes <- row.names(MixN1000X)
MixN1000X$logFC <- (MixN1X$logFC.EXPO1000X)

#Comp <- compareResultsFC(MixN1X, MixN1000X, FDRth=0.05, FCth=2^0.5, FDRceil=1e-10, logCPMth = 0, FCceil=8, title='Comparing MixN 1X to 1000X', geneLabel=TRUE)

#Comp$Scatter
```

```{r,message=FALSE,collapse=TRUE, fig.width=10, fig.height=10}
Comp <- compareResultsFCNew(MixN1X, MixN1000X, FDRth=0.05, FCth=2^0.5, FDRceil=1e-10, logCPMth=0, FCceil=2.5, 
                              title='Comparing MixN 1X to 1000X', geneLabel=TRUE, topLab=-16)

Comp$Scatter

ggsave(Comp$Scatter, filename='Data/ChronicOrganoids/Scatter.pdf', width=10, height=10)
ggsave(Comp$Scatter, filename='Data/ChronicOrganoids/Scatter.png', width=10, height=10)
```



### Neurodevelopmental Disorder (NDD) Genes 

```{r, message=FALSE,collapse=TRUE, fig.height=6}
load("Data/ASD.RData", verbose = T)
org.chronic3 <- row.names(DEAs$chronic.org)[which(DEAs$chronic.org$FDR<=0.05)]
controlNeg <- row.names(DEAs$chronic.org)[-which(DEAs$chronic.org$FDR<=0.05)]


chronic.orgUp <- row.names(DEAs$chronic.org)[which(((DEAs$chronic.org$logFC.EXPO1X)>0 & (DEAs$chronic.org$logFC.EXPO1000X)>0) & DEAs$chronic.org$FDR<=0.05)]
chronic.orgDown <- row.names(DEAs$chronic.org)[which(((DEAs$chronic.org$logFC.EXPO1X)<0 & (DEAs$chronic.org$logFC.EXPO1000X)<0) & DEAs$chronic.org$FDR<=0.05)]


#MixN <- list(organoidsDEGs=org.chronic3, organoidsDEGsUp=chronic.orgUp, organoidsDEGsDown=chronic.orgDown,NonAffectedGenes=controlNeg)
MixN <- list(organoidsDEGs=org.chronic3, NonAffectedGenes=controlNeg)

m <- overlapper::multintersect(ll = MixN, ll2 = ASD, universe = row.names(DEAs$chronic.org),two.tailed = F)

org.chronicLong<-org.chronic3
save(org.chronic,org.chronicLong,file = "Data/DEGsOrganoidsChronic.RData")
```



```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, sizeRange = c(0,15), th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicOrganoids/DotplotOrgNDD.pdf', width=10, height=6)

```

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m$prob <- round(m$prob, digits=3)
#DT::datatable(m$prob)
```

### SFARI genes
```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m <- overlapper::multintersect(ll = MixN, ll2 = SFARIgenes, universe = row.names(DEAs$chronic.org),two.tailed = F)
```

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicOrganoids/DotplotOrgSFARI.pdf', width=10, height=6)
```

### NPD genes
```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m <- overlapper::multintersect(ll = MixN, ll2 = NeuropsychiatricDiseases, universe = row.names(DEAs$chronic.org),two.tailed = F)
```

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicOrganoids/DotplotOrgNPD.pdf', width=10, height=6)
```

### NDD psychencode
```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
m <- overlapper::multintersect(ll = MixN, ll2 = PsychencodeNDD, universe = row.names(DEAs$chronic.org),two.tailed = F)
```

```{r, message=FALSE,collapse=TRUE, fig.height=6, fig.width=10}
dotplot.multintersect(m, th=0.05)

ggsave(dotplot.multintersect(m, th=0.05), filename='Data/ChronicFetal/DotplotOrgNDD.pdf', width=10, height=6)
```


### Gene expression dysregulation of the most important NDD genes

#### highest scores

```{r, message=FALSE,collapse=TRUE,fig.width=10}
geneStripPairEDCMix(SE = MixN_org_chronic,GeneSet =intersect(org.chronic, union(union(SFARIgenes$score1,SFARIgenes$score2),SFARIgenes$score3)), printExp = FALSE, SampleColors = "Default")
```

#### all scores

```{r, message=FALSE,collapse=TRUE,fig.width=15}
geneStripPairEDCMix(SE = MixN_org_chronic,GeneSet =intersect(org.chronic, unlist(SFARIgenes)), printExp = FALSE, SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.org[intersect(org.chronic, unlist(SFARIgenes)),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`


### Gene expression dysregulation of the fetal NDD-DEGs in organoids dataset

```{r, message=FALSE,collapse=TRUE}
load("Data/DEGsFetalChronic.RData", verbose = T)
geneStripPairEDCMix(SE = MixN_org_chronic,GeneSet = intersect(fetal.chronic, unlist(SFARIgenes)), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.org[intersect(fetal.chronic, unlist(SFARIgenes)),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`


### Gene expression dysregulation of the genes found in the acute expo

```{r, message=FALSE,collapse=TRUE}
load("Data/DEGsFetalAcute.RData", verbose = T)
geneStripPairEDCMix(SE = MixN_org_chronic,GeneSet = intersect(fet.acute,org.chronic), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.org[intersect(fet.acute,org.chronic),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```

 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`


```{r, message=FALSE,collapse=TRUE}
geneStripPairEDCMix(SE = MixN_org_chronic,GeneSet = c("CLSTN2", "EPHB2"), printExp = FALSE,SampleColors = "Default")
```

```{r, collapse=TRUE}
PVals <- DEAs$chronic.org[c("CLSTN2", "EPHB2"),]
PVals <- PVals[complete.cases(PVals), ]
PVals
```


 * Genes with conventional PVal > 0.05: `r row.names(PVals[PVals$PValue > 0.05,])`
 * Genes with conventional PVal < 0.05: `r row.names(PVals[PVals$PValue <= 0.05 & PVals$FDR > 0.05,])`
 * Genes with conventional FDR < 0.05: `r row.names(PVals[PVals$FDR <= 0.05,])`


### Comparison across systems 

#### more stringent DEGs
```{r, message=FALSE,collapse=TRUE}
load("Data/DEGsFetalChronic.RData", verbose = T)
load("Data/DEGsOrganoidsChronic.RData", verbose = T)

MixN <- list(OrganoidsDEGs=org.chronic, FetalAcuteDEGs=fet.acute, FetalChronicDEGs=fetal.chronic)

m<-multintersect(ll = MixN, universe = row.names(filterGenes(Total)),two.tailed = F)
dotplot.multintersect(m, th=0.05)

```

#### less stringent DEGs

```{r, message=FALSE,collapse=TRUE}
MixN <- list(OrganoidsDEGs=org.chronicLong, FetalAcuteDEGs=fet.acute, FetalChronicDEGs=fetal.chronicLong)

m<-multintersect(ll = MixN, universe = row.names(filterGenes(Total)),two.tailed = F)
dotplot.multintersect(m, th=0.05)

```



### Data praparation

For details on data filtering, normalization, batch correction and differential expression analysis, see [here](./DataPreparation/SVAcorrectedObjects.html) 

For details on the preparation of Neurodevelopmental disorder genes, see [here](./DataPreparation/ASD/DataPreparation.html) 

***

```{r child='footer.Rmd'}
```


```{r SaveSession}
sessionInfo()
```

